prov/util/src/util_ep.c

```diff
@@ -36,6 +36,38 @@
 #include <fi_enosys.h>
 #include <fi_util.h>
 
+int ofi_ep_bind_cq(struct util_ep *ep, struct util_cq *cq, uint64_t flags)
+{
+	if (flags & ~(FI_TRANSMIT | FI_RECV |
+		      FI_SELECTIVE_COMPLETION))
+		return -FI_EBADFLAGS;
+
+	if (((flags & FI_TRANSMIT) && ep->tx_cq) ||
+	    ((flags & FI_RECV) && ep->rx_cq))
+		return -FI_EINVAL;
+
+	if (flags & FI_TRANSMIT) {
+		ep->tx_cq = cq;
+		if (!(flags & FI_SELECTIVE_COMPLETION))
+			ep->tx_op_flags |= FI_COMPLETION;
+		ofi_atomic_inc32(&cq->ref);
+	}
+
+	if (flags & FI_RECV) {
+		ep->rx_cq = cq;
+		if (!(flags & FI_SELECTIVE_COMPLETION))
+			ep->rx_op_flags |= FI_COMPLETION;
+		ofi_atomic_inc32(&cq->ref);
+	}
+
+	if (flags & (FI_TRANSMIT | FI_RECV)) {
+		return fid_list_insert(&cq->ep_list,
+				       &cq->ep_list_lock,
+				       &ep->ep_fid.fid);
+	}
+
+	return FI_SUCCESS;
+}
```

Comment:
Please move the debug print statements into this call.  I think you can access the provider structure from the ep, but if not, just pass in the provider structure to the call.